# Password exploration

## PART 1: Data pre-process
### 1.1 Read password

In [1]:
import pandas as pd
import numpy as np

data = pd.read_csv('../data/38650-password-sktorrent.txt', header=None)

### 1.2 Create a table

In [2]:
import re

data.columns = ['original']
    
data['original'] = data['original'].astype(str)


Wordsegmet Package and function for preprocessing list

In [259]:
import wordsegment as ws

from wordsegment import load, segment

load()
  
def flat(l):
    for k in l:
        if not isinstance(k, (list, tuple)):
            yield k
        else:
            yield from flat(k)

### 1.3 Splitting

In [4]:
# split every characters
data["simplesplit"] = ""
# split special characters
data["specialcharsplit"] = ""
# split words but don't remain the captial characters
data["extrawordsplit"] = ""



for index, row in data.iterrows():
    # simple split for each row
    row['simplesplit'] = list(row['original'])
    row['specialcharsplit'] = re.findall(r'[0-9]+|[a-zA-Z]+|[^a-zA-Z0-9]+',row['original'])
    
    passwdstring=[]
    for i in row['specialcharsplit']:
        templist = segment(i)
        if(templist):
            passwdstring.append(templist)
        else:
            passwdstring.append(i)
    row['extrawordsplit'] = list(flat(passwdstring))

In [5]:
def insertChar(mystring, position, chartoinsert ):
    longi = len(mystring)
    mystring   =  mystring[:position] + chartoinsert + mystring[position:] 
    return mystring

In [6]:
# split words and remain the captial charaters (fix captial issues)
data["fixcapital"] = ""

for index, row in data.iterrows():
    temp = row['extrawordsplit']
    mixwithspace = " ".join(temp)
    origin = row['original']
    for i in range(0,len(mixwithspace)):
        if(mixwithspace[i]==" "):
            origin = insertChar(origin,i," ")
    row["fixcapital"] = origin

In [7]:
import pickle
with open('../data/data_1.pickle', 'wb') as f:
    pickle.dump(data, f)

## PART 2: Count frequency and Bag of words

In [8]:
import pickle

with open('../data/data_1.pickle', 'rb') as f:
    data = pickle.load(f)

 <font color='red'>TODO: need to modify this part</font> 

Combine password into a single documents

In [9]:
temp = ""
for i in data['fixcapital']:
    temp = temp +" "+i
#temp

Count frequency

In [10]:
frequency_fixcaptial = {}
for i in range(0,len(data['fixcapital'])):
    for word in data['fixcapital'][i].split():
        if word not in frequency_fixcaptial:
            frequency_fixcaptial[word] = 1
        else:
            frequency_fixcaptial[word] += 1
temp=[]
for i in range(0,len(data['specialcharsplit'])):
    temp.append(" ".join(data['specialcharsplit'][i]))

frequency = {}
for i in range(0,len(data['specialcharsplit'])):
    for word in temp[i].split():
        if word not in frequency:
            frequency[word] = 1
        else:
            frequency[word] += 1
frequency_fixcaptial = sorted(frequency_fixcaptial.items(),key = lambda item:item[1], reverse=True)
frequency = sorted(frequency.items(),key = lambda item:item[1], reverse=True)

In [11]:
#frequency
#frequency_fixcaptial

## 2.2 Bag of words

### 2.2.1 Data pre-process

In [12]:
import pickle

with open('../data/data_1.pickle', 'rb') as f:
    data = datatemp = pickle.load(f)

Remove split elements with small length

In [13]:
for index, row in datatemp.iterrows():
    for i, val in enumerate(row['extrawordsplit']):
        if len((row['extrawordsplit'])[i]) == 2:
            (row['extrawordsplit'])[i] = ''
datatemp['joinwithout2'] = ""
for index, row in datatemp.iterrows():
    row['joinwithout2'] = " ".join(row['extrawordsplit'])

In [14]:
for index, row in datatemp.iterrows():
    for i, val in enumerate(row['extrawordsplit']):
        if len((row['extrawordsplit'])[i]) == 1:
            (row['extrawordsplit'])[i] = ''
datatemp['joinwithout2n1'] = ""
for index, row in datatemp.iterrows():
    row['joinwithout2n1'] = " ".join(row['extrawordsplit'])

In [15]:
data['joinwithout2'] = datatemp['joinwithout2']
data['joinwithout2n1'] = datatemp['joinwithout2n1']

In [16]:
import pickle
with open('../data/data_2.pickle', 'wb') as f:
    pickle.dump(data, f)

### 2.2.2 Count the TF and get tf–idf matrix

In [257]:
import pickle

with open('../data/data_2.pickle', 'rb') as f:
    data = pickle.load(f)

In [260]:
texts = data['joinwithout2n1']
texts = list(flat(texts))
my_stop_words = ['']

In [94]:
texts

['   xao  ',
 'milan  ',
 'pan chart 1983',
 ' pink  123',
 'bulldog',
 'nere knu 9558285913',
 '  sek',
 '123456  ',
 'lumpy 1976',
 'simon  ',
 'kur vicky',
 'mlgqtazy ',
 'pema kuh  2834 ',
 'fer  man',
 '131619',
 'detail ist',
 'debus trol',
 'mixvibes 123',
 '161616',
 'hob lly ',
 'stano 2724',
 '19901807',
 '2925845',
 ' sen jet ',
 '565455',
 'ubopriqpjq',
 'mvl     ',
 '12082389',
 'tomas 2008',
 'jde mena ',
 '10581058',
 'kosmo 007',
 'floorball 305',
 'michal ',
 'aviat  ',
 'the encore ',
 ' per',
 ' giu   224',
 '14081982',
 'mam sarada 1984',
 'maktub',
 'vase  3962521993',
 'bul   ',
 'vandal 1982',
 'jackson',
 'partizan 001',
 'cho  ',
 '159635789',
 'katka 2224',
 ' rozen   140977',
 'falk  ',
 '  kero',
 '     ',
 'ext    ',
 'eii rjt  jrh',
 'jirka hra sky',
 'mar tuska 12365',
 ' sika 7771',
 ' jan  2028',
 'pros  182',
 'zajac 222',
 '06064991',
 'rog ippon ',
 'melina ',
 ' auch 808',
 ' jus  ',
 'gargamel 19790222',
 'market ',
 'pep ice ',
 ' esko ',
 '888 ku

In [261]:
from sklearn.feature_extraction.text import TfidfVectorizer

# change the max_df to get different results

tfidf_vectorizer = TfidfVectorizer(max_df=0.999, max_features=200000,
                                   min_df=0.001, stop_words=my_stop_words, 
                                   use_idf=True, ngram_range=(1,3))

%time tfidf_matrix = tfidf_vectorizer.fit_transform(texts)

print(tfidf_matrix.shape)

CPU times: user 262 ms, sys: 8.87 ms, total: 271 ms
Wall time: 271 ms
(38650, 63)


In [263]:
terms = tfidf_vectorizer.get_feature_names()

K-means cluster

In [264]:
from sklearn.cluster import KMeans
# num of clusters depends on dataset 
num_clusters = 10
km = KMeans(n_clusters=num_clusters)
%time km.fit(tfidf_matrix)
clusters = km.labels_.tolist()

CPU times: user 1.12 s, sys: 4.36 ms, total: 1.12 s
Wall time: 1.13 s


In [22]:
# turn the sparse matrix into matrix
temp = tfidf_matrix.todense()
#temp

Gaussian Mixture model

In [23]:
from sklearn.mixture import GaussianMixture
%time gmm = GaussianMixture(n_components=10).fit(temp) 
labels1 = gmm.predict(temp)

CPU times: user 14.2 s, sys: 459 ms, total: 14.7 s
Wall time: 2.46 s


Dirichlet Process and GMM

In [24]:
from sklearn.mixture import BayesianGaussianMixture
#gmm = BayesianGaussianMixture(n_components=100).fit(temp) 
dpgmm = BayesianGaussianMixture(n_components=10,covariance_type='full').fit(temp)
labels2 = dpgmm.predict(temp)

DBSCAN (about 2 min)

In [25]:
from sklearn.cluster import DBSCAN
%time labels3 = DBSCAN().fit_predict(temp)

CPU times: user 2min 19s, sys: 22.1 s, total: 2min 41s
Wall time: 2min 44s


In [26]:
import pandas as pd
# Get the TF-IDF matrix
df_tfidf_matrix = pd.DataFrame(tfidf_matrix.toarray())

<font color='red'>TODO: need to solve the username issue</font> 

In [27]:
import re

data1 = pd.read_csv('../data/38650-username-sktorrent.txt', encoding='ISO-8859-1', header=None)

data1.columns = ['original']
    
data1['original'] = data1['original'].astype(str)

data1 = data1['original'][0:38650]
#data1

Comparison of different algorithm

In [28]:
news = {'username': data1,'text': texts,'K_means_cluster': clusters,'GMM':labels1,'DPGMM':labels2,'DBSCAN':labels3}

frame = pd.DataFrame(news, columns = ['username','text','K_means_cluster','GMM','DPGMM','DBSCAN'])

In [29]:
frame.head()

username            text  K_means_cluster  GMM  DPGMM  DBSCAN
0   otlak33           xao                  0    0      0       0
1  bocko202         milan                  0    1      3       1
2  Finochio  pan chart 1983                0    0      0       0
3   Marobod       pink  123                2    1      3       2
4     tomos         bulldog                0    0      0       0

In [30]:
import pickle

with open('../data/data_clusters.pickle', 'wb') as f:
    pickle.dump(frame, f)

In [31]:
frame['DPGMM'].value_counts().head()

0    34283
3     1395
7      438
6      436
4      399
Name: DPGMM, dtype: int64

Here we can see there is a "big" class dominating the clusters which is not what we expected. The reason might be the passwords' weeker relationship and the first class might contain passwords without obvious relationship.

In [32]:
print("Top terms per cluster:")
print()
order_centroids = km.cluster_centers_.argsort()[:, ::-1]
for i in range(num_clusters):
    print("Cluster %d words:" % i, end='')
    for ind in order_centroids[i, :6]:
        print(' %s' % terms[ind].split(' '), end=',')
    print()
    print()

Top terms per cluster:

Cluster 0 words: ['123456'], ['koko'], ['111'], ['david'], ['007'], ['lukas'],

Cluster 1 words: ['cek'], ['123'], ['1997'], ['filip'], ['12345'], ['666'],

Cluster 2 words: ['123'], ['koko'], ['456'], ['lol'], ['pet'], ['jakub'],

Cluster 3 words: ['666'], ['the'], ['filip'], ['456'], ['marek'], ['koko'],

Cluster 4 words: ['martin'], ['123'], ['1996'], ['1987'], ['the'], ['1986'],

Cluster 5 words: ['1234'], ['david'], ['koko'], ['martin'], ['dominik'], ['sky'],

Cluster 6 words: ['heslo'], ['123'], ['torrent'], ['123456'], ['2014'], ['1986'],

Cluster 7 words: ['torrent'], ['123'], ['333'], ['007'], ['2014'], ['filip'],

Cluster 8 words: ['michal'], ['12345'], ['ska'], ['1987'], ['1986'], ['321'],

Cluster 9 words: ['jan'], ['123'], ['kova'], ['cik'], ['enka'], ['ova'],



Nearly every class contians 123 which is used widely for password and easy to remember.

In [93]:
len(order_centroids[1])

63

The results are quiet good for the small size classes. For example: the cluster 1 contains 123

In [33]:
def showthehead(data,clusters_data,ind):
    index = []
    for i in range(0,len(clusters)-1):
        if clusters[i]==ind:
            index.append(i)
    print(data['original'][index].head())

We can see these clusters all have some common things, it might because the clusters algorithm is based on the distance between the matrix.

In [34]:
print("Top terms per cluster:")

showthehead(data,clusters,1)

Top terms per cluster:
1697     1dinecek
1931     kundicek
2018      marecek
2492     filipcek
2710    misacek86
Name: original, dtype: object


In [35]:
index = []
for i in range(0,len(labels1)-1):
    if labels1[i]==3:
        index.append(i)
data['original'][index].head()

1290        Ali123456789
1483    setobox123456789
1612       voda123456789
1957      Peter123456789
2379      123456789matko
Name: original, dtype: object

In [36]:
index = []
for i in range(0,len(labels2)-1):
    if labels2[i]==3:
        index.append(i)
data['original'][index].head()

1       milanek22
3      pipinka123
13        ferbman
17    mixvibes123
68        pepicek
Name: original, dtype: object

In [37]:
index = []
for i in range(0,len(labels3)-1):
    if labels3[i]==3:
        index.append(i)
data['original'][index].head()
#sek

6         blbysek
442    Ondrasek98
492     matysek45
836      mikisek5
930      dandysek
Name: original, dtype: object

...continuous

In [38]:
import pickle

with open('../data/data_clusters.pickle', 'rb') as f:
    passwd_class = pickle.load(f)

## Markov Chain

In [39]:
import pickle

with open('../data/data_2.pickle', 'rb') as f:
    data = pickle.load(f)

Try to get a transition matrix

In [196]:
tranlist = []
count = 0
for passwd in data['simplesplit']:
    for i in range(len(passwd)-1):
        temp = passwd[i] + passwd[i+1]
        if temp in tranlist:
            count = count + 1
        else:
            tranlist.append(temp)
tranlist.index('V2')


3806

In [245]:
import re
alphabet = ''
for index, row in data.iterrows():
    alphabet = alphabet + row['original']


{' ',
 '!',
 '#',
 '$',
 '%',
 '&',
 '(',
 ')',
 '*',
 '+',
 '-',
 '.',
 '/',
 '0',
 '1',
 '2',
 '3',
 '4',
 '5',
 '6',
 '7',
 '8',
 '9',
 ':',
 '<',
 '=',
 '>',
 '?',
 '@',
 'A',
 'B',
 'C',
 'D',
 'E',
 'F',
 'G',
 'H',
 'I',
 'J',
 'K',
 'L',
 'M',
 'N',
 'O',
 'P',
 'Q',
 'R',
 'S',
 'T',
 'U',
 'V',
 'W',
 'X',
 'Y',
 'Z',
 '[',
 ']',
 '^',
 '_',
 'a',
 'b',
 'c',
 'd',
 'e',
 'f',
 'g',
 'h',
 'i',
 'j',
 'k',
 'l',
 'm',
 'n',
 'o',
 'p',
 'q',
 'r',
 's',
 't',
 'u',
 'v',
 'w',
 'x',
 'y',
 'z',
 '{',
 '©',
 'µ',
 '·',
 '¹',
 '»',
 '¾',
 'á',
 'è',
 'é',
 'ì',
 'í',
 'ï',
 'ø',
 'ù',
 'ú',
 'ý'}

In [ ]:
alphabet = list(set(alphabet))

for i in range(len(alphabet)-1):
    for j in range(len(alphabet)-1):
        tranlist.index(alphabet[i] + alphabet[j])

In [ ]:
for i in range(len(tranlist)):
    tranlist.index(tranlist[i][::-1])

In [255]:
alphabet[i] + alphabet[i+1]

'ú¹'

In [ ]:
indptr = np.array([0, 2, 3, 6])
indices = np.array([0, 2, 2, 0, 1, 2])
data = np.array([1, 2, 3, 4, 5, 6])
sparse.csr_matrix((data, indices, indptr), shape=(3, 3)).toarray()

In [256]:
i = 1
try:
    tranlist.index(alphabet[i] + alphabet[j])
except:
    

ValueError: 'ú¹' is not in list

In [41]:
tranlist = []
for passwd in data['simplesplit']:
    for i in range(len(passwd)-1):
        temp = passwd[i]+passwd[i+1]
        tranlist.append(temp)
            



In [216]:
count.items()
set(count)
sum(count.values())

4722

a = {} 
for i in tranlist:
    if tranlist.count(i)>1: 
        a[i] = tranlist.count(i) 
a = sorted(a.items(), key=lambda item:item[0])

## LDA

In [4]:
import pyLDAvis
import pyLDAvis.sklearn
pyLDAvis.enable_notebook()

In [5]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.decomposition import LatentDirichletAllocation

In [6]:
import pickle

with open('../data/data_2.pickle', 'rb') as f:
    data = pickle.load(f)

In [7]:
docs_raw = list(data['fixcapital'])
print(len(docs_raw))

38650


### Try LDA with words only

In [165]:
def LDA_words(docs_raw,n_topics):
    tf_vectorizer = CountVectorizer(strip_accents = 'unicode',
                                    stop_words = 'english',
                                    lowercase = True,
                                    # token choose the words with more than 4 length and get rid of the speical words
                                    token_pattern = r'\b[a-zA-Z]{3,}\b',
                                    max_df = 0.5, 
                                    min_df = 10)
    dtm_tf = tf_vectorizer.fit_transform(docs_raw)
    tfidf_vectorizer = TfidfVectorizer(**tf_vectorizer.get_params())
    dtm_tfidf = tfidf_vectorizer.fit_transform(docs_raw)
    # for TF DTM
    lda_tf = LatentDirichletAllocation(n_components=n_topics, random_state=0)
    %time lda_tf.fit(dtm_tf)
    # for TFIDF DTM
    lda_tfidf = LatentDirichletAllocation(n_components=n_topics, random_state=0)
    %time lda_tfidf.fit(dtm_tfidf)
    return(lda_tf, dtm_tf, tf_vectorizer)

def LDA_words_numbers(docs_raw,n_topics):
    tf_vectorizer = CountVectorizer(strip_accents = 'unicode',
                                    stop_words = 'english',
                                    lowercase = True,
                                    # token choose the words with more than 4 length and get rid of the speical words
                                    token_pattern = r'\b[a-zA-Z0-9]{3,}\b',
                                    max_df = 0.5, 
                                    min_df = 10)
    dtm_tf = tf_vectorizer.fit_transform(docs_raw)
    tfidf_vectorizer = TfidfVectorizer(**tf_vectorizer.get_params())
    dtm_tfidf = tfidf_vectorizer.fit_transform(docs_raw)
    # for TF DTM
    lda_tf = LatentDirichletAllocation(n_components=n_topics, random_state=0)
    %time lda_tf.fit(dtm_tf)
    # for TFIDF DTM
    lda_tfidf = LatentDirichletAllocation(n_components=n_topics, random_state=0)
    %time lda_tfidf.fit(dtm_tfidf)
    return(lda_tf, dtm_tf, tf_vectorizer)



#### It takes about 1 min

In [14]:
# lda_tf_withwords
lda_tf_ww, dtm_tf_ww, tf_vectorizer_ww = LDA_words(docs_raw,10)
pyLDAvis.sklearn.prepare(lda_tf_ww, dtm_tf_ww, tf_vectorizer_ww)

CPU times: user 18.1 s, sys: 21.2 ms, total: 18.1 s
Wall time: 17.3 s
CPU times: user 17.1 s, sys: 35.5 ms, total: 17.2 s
Wall time: 16.4 s


PreparedData(topic_coordinates=              x         y  topics  cluster       Freq
topic                                                
3      0.044029 -0.284039       1        1  10.760423
0     -0.254793  0.245360       2        1  10.741163
9     -0.005582 -0.119118       3        1  10.405213
6      0.202879  0.153181       4        1  10.223639
8     -0.206028 -0.031013       5        1  10.159828
1      0.207196  0.115698       6        1  10.106019
4     -0.026947 -0.031870       7        1  10.071468
2     -0.067980 -0.098069       8        1   9.504106
5      0.003828 -0.002370       9        1   9.268863
7      0.103398  0.052241      10        1   8.759278, topic_info=     Category        Freq     Term       Total  loglift  logprob
term                                                            
290   Default  133.000000   martin  133.000000  30.0000  30.0000
516   Default  125.000000  torrent  125.000000  29.0000  29.0000
127   Default  123.000000    heslo  123.000000  28.0000  28.0000
150   Default  118.000000      jan  118.000000  27.0000  27.0000
141   Default  103.000000      ink  103.000000  26.0000  26.0000
204   Default  109.000000     koko  109.000000  25.0000  25.0000
54    Default  104.000000      cek  104.000000  24.0000  24.0000
72    Default  100.000000    david  100.000000  23.0000  23.0000
262   Default   93.000000    lukas   93.000000  22.0000  22.0000
138   Default   84.000000      ice   84.000000  21.0000  21.0000
281   Default   86.000000    marek   86.000000  20.0000  20.0000
148   Default   83.000000    jakub   83.000000  19.0000  19.0000
470   Default   83.000000      ska   83.000000  18.0000  18.0000
513   Default   81.000000    tomas   81.000000  17.0000  17.0000
395   Default   75.000000    peter   75.000000  16.0000  16.0000
308   Default   74.000000   michal   74.000000  15.0000  15.0000
462   Default   73.000000      sek   73.000000  14.0000  14.0000
373   Default   67.000000      ova   67.000000  13.0000  13.0000
277   Default   67.000000      man   67.000000  12.0000  12.0000
7     Default   63.000000     adam   63.000000  11.0000  11.0000
250   Default   55.000000      lol   55.000000  10.0000  10.0000
316   Default   53.000000    milan   53.000000   9.0000   9.0000
248   Default   57.000000     link   57.000000   8.0000   8.0000
394   Default   55.000000      pet   55.000000   7.0000   7.0000
355   Default   54.000000      nik   54.000000   6.0000   6.0000
353   Default   50.000000     nice   50.000000   5.0000   5.0000
245   Default   52.000000      lik   52.000000   4.0000   4.0000
278   Default   51.000000      mar   51.000000   3.0000   3.0000
104   Default   51.000000     enka   51.000000   2.0000   2.0000
354   Default   47.000000     nick   47.000000   1.0000   1.0000
...       ...         ...      ...         ...      ...      ...
141   Topic10  102.430099      ink  103.279525   2.4268  -2.2715
138   Topic10   83.925193      ice   84.774617   2.4250  -2.4708
316   Topic10   52.358002    milan   53.207423   2.4190  -2.9426
385   Topic10   36.030140      pat   36.879563   2.4118  -3.3163
482   Topic10   32.764565      som   33.613990   2.4095  -3.4113
345   Topic10   31.676038      nec   32.525466   2.4086  -3.4451
349   Topic10   28.410471      net   29.259895   2.4056  -3.5539
310   Topic10   27.321948      mik   28.171371   2.4044  -3.5930
259   Topic10   25.144902      luc   25.994323   2.4018  -3.6760
147   Topic10   24.056373      jak   24.905799   2.4004  -3.7203
438   Topic10   21.879331   robert   22.728751   2.3970  -3.8151
291   Topic10   20.790807  martina   21.640227   2.3950  -3.8662
132   Topic10   20.790804      hon   21.640227   2.3950  -3.8662
495   Topic10   18.613760      str   19.463180   2.3904  -3.9768
165   Topic10   18.613751      jka   19.463178   2.3904  -3.9768
558   Topic10   16.436712      zik   17.286132   2.3847  -4.1012
377   Topic10   16.436712     pali   17.286132   2.3847  -4.1012
5     Topic10   16.436711     acer   17.286131   2.3847  -4

As we can see the "heslo" which means password shows a really high frequency in the topic 1, and also if we pay attention to other topics we can find torrent(a part of the website name). If we use different password dataset, can we always find that keywords? If so, maybe we can use LDA to guess where the data from automatically. In addtion, maybe we can find the common password keyword by using LDA in different datasets. However, as each password contains few information and have a week relationship, we cannot get what we expected -- the first topic is the people's name, second one is common passwords and so on.

In [19]:
# Take about 1 min
# lda_tf_withwords_numbers
lda_tf_wwn, dtm_tf_wwn, tf_vectorizer_wwn = LDA_words_numbers(docs_raw,10)
pyLDAvis.sklearn.prepare(lda_tf_wwn, dtm_tf_wwn, tf_vectorizer_wwn)

/Users/xihajun/anaconda3/lib/python3.7/site-packages/sklearn/feature_extraction/text.py:1567: UserWarning: Only (<class 'numpy.float64'>, <class 'numpy.float32'>, <class 'numpy.float16'>) 'dtype' should be used. <class 'numpy.int64'> 'dtype' will be converted to np.float64.
  UserWarning)


CPU times: user 18.6 s, sys: 24.4 ms, total: 18.7 s
Wall time: 17.9 s
CPU times: user 17.6 s, sys: 21.6 ms, total: 17.6 s
Wall time: 16.8 s


/Users/xihajun/anaconda3/lib/python3.7/site-packages/pyLDAvis/_prepare.py:257: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  return pd.concat([default_term_info] + list(topic_dfs))


PreparedData(topic_coordinates=              x         y  topics  cluster       Freq
topic                                                
9      0.411023  0.065278       1        1  12.314293
5     -0.092114 -0.170043       2        1  11.124532
4     -0.117295  0.273186       3        1  10.183388
1      0.017655 -0.139334       4        1  10.000175
7     -0.102958  0.226239       5        1   9.708437
8     -0.056792 -0.096331       6        1   9.637800
0     -0.018505  0.068639       7        1   9.489637
3     -0.061269 -0.098359       8        1   9.301952
2      0.076112 -0.016259       9        1   9.220286
6     -0.055856 -0.113017      10        1   9.019500, topic_info=     Category        Freq       Term       Total  loglift  logprob
term                                                              
13    Default  897.000000        123  897.000000  30.0000  30.0000
92    Default  145.000000        666  145.000000  29.0000  29.0000
391   Default  143.000000     martin  143.000000  28.0000  28.0000
617   Default  130.000000    torrent  130.000000  27.0000  27.0000
228   Default  130.000000      heslo  130.000000  26.0000  26.0000
15    Default  128.000000       1234  128.000000  25.0000  25.0000
251   Default  123.000000        jan  123.000000  24.0000  24.0000
305   Default  114.000000       koko  114.000000  23.0000  23.0000
17    Default  109.000000     123456  109.000000  22.0000  22.0000
155   Default  104.000000        cek  104.000000  21.0000  21.0000
363   Default   98.000000      lukas   98.000000  20.0000  20.0000
7     Default   98.000000        111   98.000000  19.0000  19.0000
571   Default   86.000000        ska   86.000000  18.0000  18.0000
614   Default   87.000000      tomas   87.000000  17.0000  17.0000
563   Default   84.000000        sek   84.000000  16.0000  16.0000
173   Default   92.000000      david   92.000000  15.0000  15.0000
3     Default   87.000000        007   87.000000  14.0000  14.0000
496   Default   80.000000      peter   80.000000  13.0000  13.0000
242   Default   85.000000        ink   85.000000  12.0000  12.0000
239   Default   77.000000        ice   77.000000  11.0000  11.0000
382   Default   82.000000      marek   82.000000  10.0000  10.0000
474   Default   73.000000        ova   73.000000   9.0000   9.0000
16    Default   73.000000      12345   73.000000   8.0000   8.0000
249   Default   74.000000      jakub   74.000000   7.0000   7.0000
18    Default   65.000000  123456789   65.000000   6.0000   6.0000
108   Default   66.000000       adam   66.000000   5.0000   5.0000
409   Default   70.000000     michal   70.000000   4.0000   4.0000
495   Default   60.000000        pet   60.000000   3.0000   3.0000
378   Default   63.000000        man   63.000000   2.0000   2.0000
346   Default   55.000000        lik   55.000000   1.0000   1.0000
...       ...         ...        ...         ...      ...      ...
563   Topic10   83.151471        sek   84.014402   2.3955  -2.8069
437   Topic10   38.513349     monika   39.376279   2.3836  -3.5765
57    Topic10   38.513348       1995   39.376280   2.3836  -3.5765
532   Topic10   37.475256       rent   38.338184   2.3830  -3.6039
56    Topic10   37.475249       1994   38.338184   2.3830  -3.6039
256   Topic10   36.437153        jar   37.300088   2.3824  -3.6319
51    Topic10   35.399059       1989   36.261992   2.3817  -3.6609
79    Topic10   33.322869        222   34.185800   2.3802  -3.7213
121   Topic10   32.284772       asdf   33.147704   2.3794  -3.7529
593   Topic10   31.246682       star   32.109609   2.3785  -3.7856
640   Topic10   27.094294       vice   27.957225   2.3744  -3.9282
578   Topic10   27.094290       slav   27.957225   2.3744  -3.9282
168   Topic10   27.094288        dan   27.957225   2.3744  -3.9282
103   Topic10   26.056198        abc   26.919129   2.3732  -3.9673
272   Topic10   26.056196        jul   26.919129   2.3732  -3.9673
52    Topic10   26.056194       1990   26.919129   2.3732  -3.9673
41    Topic10   22.941910 

Here we can see the most salient terms namely 123, 666, martin, torrent, heslo, and some topics' distances might be optimalized when we only require 5 topics.

In [20]:
# Take about 1 min
# lda_tf_withwords_numbers
lda_tf_wwn, dtm_tf_wwn, tf_vectorizer_wwn = LDA_words_numbers(docs_raw,5)
pyLDAvis.sklearn.prepare(lda_tf_wwn, dtm_tf_wwn, tf_vectorizer_wwn)

CPU times: user 18 s, sys: 20.4 ms, total: 18 s
Wall time: 18 s
CPU times: user 16.8 s, sys: 28 ms, total: 16.8 s
Wall time: 16.8 s


PreparedData(topic_coordinates=              x         y  topics  cluster       Freq
topic                                                
0      0.330121 -0.103736       1        1  21.301821
4     -0.221044 -0.330120       2        1  20.351252
3     -0.114960  0.159050       3        1  19.595099
1     -0.119893  0.223899       4        1  19.419517
2      0.125777  0.050906       5        1  19.332311, topic_info=     Category        Freq       Term       Total  loglift  logprob
term                                                              
13    Default  970.000000        123  970.000000  30.0000  30.0000
92    Default  146.000000        666  146.000000  29.0000  29.0000
391   Default  136.000000     martin  136.000000  28.0000  28.0000
228   Default  132.000000      heslo  132.000000  27.0000  27.0000
617   Default  131.000000    torrent  131.000000  26.0000  26.0000
251   Default  123.000000        jan  123.000000  25.0000  25.0000
15    Default  124.000000       1234  124.000000  24.0000  24.0000
305   Default  111.000000       koko  111.000000  23.0000  23.0000
17    Default  106.000000     123456  106.000000  22.0000  22.0000
155   Default  104.000000        cek  104.000000  21.0000  21.0000
363   Default   96.000000      lukas   96.000000  20.0000  20.0000
7     Default   98.000000        111   98.000000  19.0000  19.0000
173   Default   97.000000      david   97.000000  18.0000  18.0000
3     Default   93.000000        007   93.000000  17.0000  17.0000
382   Default   90.000000      marek   90.000000  16.0000  16.0000
614   Default   88.000000      tomas   88.000000  15.0000  15.0000
563   Default   79.000000        sek   79.000000  14.0000  14.0000
571   Default   79.000000        ska   79.000000  13.0000  13.0000
249   Default   79.000000      jakub   79.000000  12.0000  12.0000
409   Default   79.000000     michal   79.000000  11.0000  11.0000
474   Default   71.000000        ova   71.000000  10.0000  10.0000
496   Default   74.000000      peter   74.000000   9.0000   9.0000
378   Default   70.000000        man   70.000000   8.0000   8.0000
239   Default   71.000000        ice   71.000000   7.0000   7.0000
16    Default   68.000000      12345   68.000000   6.0000   6.0000
108   Default   64.000000       adam   64.000000   5.0000   5.0000
18    Default   62.000000  123456789   62.000000   4.0000   4.0000
349   Default   60.000000       link   60.000000   3.0000   3.0000
495   Default   57.000000        pet   57.000000   2.0000   2.0000
242   Default   91.000000        ink   91.000000   1.0000   1.0000
...       ...         ...        ...         ...      ...      ...
617    Topic5  130.525798    torrent  131.291060   1.6375  -3.1184
363    Topic5   95.248411      lukas   96.013376   1.6354  -3.4334
563    Topic5   78.589077        sek   79.354481   1.6337  -3.6257
474    Topic5   70.750162        ova   71.514996   1.6326  -3.7308
454    Topic5   48.212038       nice   48.976482   1.6277  -4.1143
96     Topic5   48.211380        777   48.976471   1.6276  -4.1143
212    Topic5   48.211090      filip   48.976455   1.6276  -4.1143
487    Topic5   44.292008     patrik   45.056749   1.6263  -4.1991
25     Topic5   44.291709        159   45.056738   1.6263  -4.1991
577    Topic5   41.351235        sky   42.116903   1.6250  -4.2678
265    Topic5   36.452304      jirka   37.217255   1.6226  -4.3939
437    Topic5   36.451931     monika   37.217241   1.6226  -4.3939
206    Topic5   36.451256       erik   37.217198   1.6226  -4.3939
464    Topic5   32.533597      nokia   33.297536   1.6202  -4.5077
358    Topic5   32.532783        lub   33.297506   1.6202  -4.5077
1      Topic5   31.552207        001   32.317571   1.6194  -4.5383
121    Topic5   30.572031       asdf   31.337615   1.6187  -4.5698
44     Topic5   29.592408       1982   30.357697   1.6179  -4.6024
583    Topic5   29.592272        som   30.357700   1.6179  -4.6024
625    Topic5   28.613334        use   29.377789   1.6170  -4.6361
384    Topic5   27.632882   

Above sumulation, we cannot get too much useful information we want. But if we get the label of the password strength things might be different?

## Password Strength

In [75]:
import pickle

with open('../data/data_2.pickle', 'rb') as f:
    data = pickle.load(f)

In [96]:
from zxcvbn import zxcvbn
# Test
results = zxcvbn('1234567')
print(results)

{'password': '1234567', 'guesses': Decimal('10'), 'guesses_log10': 1.0, 'sequence': [{'pattern': 'dictionary', 'i': 0, 'j': 6, 'token': '1234567', 'matched_word': '1234567', 'rank': 9, 'dictionary_name': 'passwords', 'reversed': False, 'l33t': False, 'base_guesses': 9, 'uppercase_variations': 1, 'l33t_variations': 1, 'guesses': 9, 'guesses_log10': 0.9542425094393249}], 'calc_time': datetime.timedelta(microseconds=1015), 'crack_times_seconds': {'online_throttling_100_per_hour': Decimal('360.0000000000000199840144433'), 'online_no_throttling_10_per_second': Decimal('1'), 'offline_slow_hashing_1e4_per_second': Decimal('0.001'), 'offline_fast_hashing_1e10_per_second': Decimal('1E-9')}, 'crack_times_display': {'online_throttling_100_per_hour': '6 minutes', 'online_no_throttling_10_per_second': '1 second', 'offline_slow_hashing_1e4_per_second': 'less than a second', 'offline_fast_hashing_1e10_per_second': 'less than a second'}, 'score': 0, 'feedback': {'warning': 'This is a top-10 common pas

In [97]:
for i in range(len(results['sequence'])):
    for key, val in results['sequence'][i].items():
        if key == 'pattern':
            if val == 'dictionary':
                print(results['sequence'][i]['dictionary_name'])
            elif val == 'sequence':
                print(results['sequence'][i]['sequence_name'])
            else:
                print(val)

passwords


Take about 30s

In [98]:
data['password_strength'] = ''
data['fixcap_with_pswd_strength'] = ''
for index, row in data.iterrows():
    # simple split for each row
    
    results = zxcvbn(row['original'])
    temp = results['score']
    if temp == 0:
        password_score = 'Weakest'
    elif temp == 1:
        password_score = 'Weak'
    elif temp == 2:
        password_score = 'Medium'
    elif temp == 3:
        password_score = 'Strong'
    else:
        password_score = 'Strongest'
    row['password_strength'] = password_score
   # data['fixcap_with_pswd_strength'] = 
    



### Strength only

In [99]:
docs_raw = list(data['fixcapital']+" "+data['password_strength'])
print(len(docs_raw))

38650


In [80]:
tf_vectorizer = CountVectorizer(strip_accents = 'unicode',
                                stop_words = 'english',
                                lowercase = True,
                                # token choose the words with more than 4 length and get rid of the speical words
                                token_pattern = r'\b[a-zA-Z0-9]{3,}\b',
                                max_df = 0.5, 
                                min_df = 10)
dtm_tf = tf_vectorizer.fit_transform(docs_raw)
print(dtm_tf.shape)

(38650, 668)


In [81]:
tfidf_vectorizer = TfidfVectorizer(**tf_vectorizer.get_params())
dtm_tfidf = tfidf_vectorizer.fit_transform(docs_raw)
print(dtm_tfidf.shape)

(38650, 668)


In [82]:
# for TF DTM
lda_tf = LatentDirichletAllocation(n_components=10, random_state=0)
%time lda_tf.fit(dtm_tf)
# for TFIDF DTM
lda_tfidf = LatentDirichletAllocation(n_components=10, random_state=0)
%time lda_tfidf.fit(dtm_tfidf)

CPU times: user 35.3 s, sys: 50.4 ms, total: 35.4 s
Wall time: 24.8 s
CPU times: user 31.6 s, sys: 12.9 ms, total: 31.6 s
Wall time: 21 s


LatentDirichletAllocation(batch_size=128, doc_topic_prior=None,
             evaluate_every=-1, learning_decay=0.7,
             learning_method='batch', learning_offset=10.0,
             max_doc_update_iter=100, max_iter=10, mean_change_tol=0.001,
             n_components=10, n_jobs=None, n_topics=None, perp_tol=0.1,
             random_state=0, topic_word_prior=None,
             total_samples=1000000.0, verbose=0)

In [83]:
pyLDAvis.sklearn.prepare(lda_tf, dtm_tf, tf_vectorizer)

PreparedData(topic_coordinates=              x         y  topics  cluster       Freq
topic                                                
8     -0.050710  0.323659       1        1  17.862302
6     -0.331314 -0.166344       2        1  14.495619
3      0.298975 -0.113698       3        1  14.087279
7     -0.028863  0.249690       4        1   9.589913
0     -0.282852 -0.131606       5        1   8.572762
2      0.239983 -0.088836       6        1   7.526557
1      0.233605 -0.077404       7        1   7.210191
4     -0.271396 -0.131233       8        1   7.037376
5      0.236429 -0.089142       9        1   6.826765
9     -0.043857  0.224913      10        1   6.791236, topic_info=     Category         Freq       Term        Total  loglift  logprob
term                                                                
598   Default  9402.000000     strong  9402.000000  30.0000  30.0000
656   Default  9262.000000       weak  9262.000000  29.0000  29.0000
404   Default  9093.000000     medium  9093.000000  28.0000  28.0000
599   Default  3341.000000  strongest  3341.000000  27.0000  27.0000
657   Default  1433.000000    weakest  1433.000000  26.0000  26.0000
13    Default  1356.000000        123  1356.000000  25.0000  25.0000
92    Default   218.000000        666   218.000000  24.0000  24.0000
251   Default   192.000000        jan   192.000000  23.0000  23.0000
15    Default   190.000000       1234   190.000000  22.0000  22.0000
305   Default   183.000000       koko   183.000000  21.0000  21.0000
155   Default   180.000000        cek   180.000000  20.0000  20.0000
391   Default   183.000000     martin   183.000000  19.0000  19.0000
173   Default   161.000000      david   161.000000  18.0000  18.0000
7     Default   156.000000        111   156.000000  17.0000  17.0000
363   Default   142.000000      lukas   142.000000  16.0000  16.0000
572   Default   135.000000        ska   135.000000  15.0000  15.0000
617   Default   137.000000      tomas   137.000000  14.0000  14.0000
17    Default   141.000000     123456   141.000000  13.0000  13.0000
382   Default   133.000000      marek   133.000000  12.0000  12.0000
228   Default   145.000000      heslo   145.000000  11.0000  11.0000
497   Default   127.000000      peter   127.000000  10.0000  10.0000
620   Default   144.000000    torrent   144.000000   9.0000   9.0000
249   Default   126.000000      jakub   126.000000   8.0000   8.0000
16    Default   111.000000      12345   111.000000   7.0000   7.0000
108   Default   111.000000       adam   111.000000   6.0000   6.0000
378   Default   113.000000        man   113.000000   5.0000   5.0000
18    Default   102.000000  123456789   102.000000   4.0000   4.0000
410   Default   105.000000     michal   105.000000   3.0000   3.0000
349   Default    96.000000       link    96.000000   2.0000   2.0000
351   Default    85.000000        lol    85.000000   1.0000   1.0000
...       ...          ...        ...          ...      ...      ...
16    Topic10   110.301745      12345   111.382983   2.6798  -3.5025
351   Topic10    84.650174        lol    85.731406   2.6768  -3.7672
82    Topic10    81.632339        321    82.713573   2.6764  -3.8035
96    Topic10    74.087764        777    75.168992   2.6750  -3.9004
548   Topic10    71.069934      roman    72.151159   2.6744  -3.9420
237   Topic10    66.543178        hov    67.624410   2.6734  -4.0078
1     Topic10    48.436177        001    49.517416   2.6675  -4.3254
202   Topic10    46.927273       elka    48.008497   2.6668  -4.3571
46    Topic10    45.418358       1984    46.499583   2.6660  -4.3898
512   Topic10    45.418357        pop    46.499584   2.6660  -4.3898
584   Topic10    45.418354        som    46.499583   2.6660  -4.3898
44    Topic10    45.418348       1982    46.499583   2.6660  -4.3898
168   Topic10    39.382685        dan    40.463918   2.6625  -4.5324
472   Topic10    37.873772      ondra    38.955001   2.6614  -4.5714
61    Topic10    36.364846       1999    37.446086   2.6602  -4.61

### Strength plus feedback

Take about 20s

In [173]:
data['feedback'] = ''
for index, row in data.iterrows():
    # simple split for each row
    
    results = zxcvbn(row['original'])
    row['feedback'] = " ".join(results['feedback']['suggestions'])
    #row['feedback'] = results['feedback']['warning']+" ".join(results['feedback']['suggestions'])
    

In [61]:
docs_raw = list(data['fixcapital']+" "+data['password_strength']+" "+data['feedback'])
print(len(docs_raw))

38650


In [62]:
tf_vectorizer = CountVectorizer(strip_accents = 'unicode',
                                stop_words = 'english',
                                lowercase = True,
                                # token choose the words with more than 4 length and get rid of the speical words
                                token_pattern = r'\b[a-zA-Z0-9]{3,}\b',
                                max_df = 0.5, 
                                min_df = 10)
dtm_tf = tf_vectorizer.fit_transform(docs_raw)
print(dtm_tf.shape)

(38650, 694)


In [63]:
tfidf_vectorizer = TfidfVectorizer(**tf_vectorizer.get_params())
dtm_tfidf = tfidf_vectorizer.fit_transform(docs_raw)
print(dtm_tfidf.shape)

(38650, 694)


In [64]:
# for TF DTM
lda_tf = LatentDirichletAllocation(n_components=10, random_state=0)
%time lda_tf.fit(dtm_tf)
# for TFIDF DTM
lda_tfidf = LatentDirichletAllocation(n_components=10, random_state=0)
%time lda_tfidf.fit(dtm_tfidf)

CPU times: user 40.1 s, sys: 74.3 ms, total: 40.2 s
Wall time: 29.6 s
CPU times: user 33 s, sys: 26.9 ms, total: 33 s
Wall time: 22.4 s


LatentDirichletAllocation(batch_size=128, doc_topic_prior=None,
             evaluate_every=-1, learning_decay=0.7,
             learning_method='batch', learning_offset=10.0,
             max_doc_update_iter=100, max_iter=10, mean_change_tol=0.001,
             n_components=10, n_jobs=None, n_topics=None, perp_tol=0.1,
             random_state=0, topic_word_prior=None,
             total_samples=1000000.0, verbose=0)

In [65]:
pyLDAvis.sklearn.prepare(lda_tf, dtm_tf, tf_vectorizer)

PreparedData(topic_coordinates=              x         y  topics  cluster       Freq
topic                                                
5      0.227444 -0.032035       1        1  23.344818
1     -0.244706  0.314190       2        1  12.735087
7     -0.174878 -0.242247       3        1  10.155817
6      0.273335  0.062052       4        1   9.474384
4      0.094692 -0.095697       5        1   9.124867
8      0.288877  0.067877       6        1   8.862425
2     -0.062086  0.237867       7        1   7.363535
3      0.069625 -0.097504       8        1   6.924284
9     -0.289154  0.032353       9        1   6.753364
0     -0.183149 -0.246855      10        1   5.261417, topic_info=     Category          Freq            Term         Total  loglift  logprob
term                                                                       
620   Default  10076.000000          strong  10076.000000  30.0000  30.0000
681   Default  10794.000000            weak  10794.000000  29.0000  29.0000
420   Default   9370.000000          medium   9370.000000  28.0000  28.0000
621   Default   3320.000000       strongest   3320.000000  27.0000  27.0000
129   Default   5157.000000           avoid   5157.000000  26.0000  26.0000
689   Default   3478.000000           years   3478.000000  25.0000  25.0000
124   Default   3333.000000      associated   3333.000000  24.0000  24.0000
178   Default   3188.000000           dates   3188.000000  23.0000  23.0000
553   Default   1543.000000        repeated   1543.000000  22.0000  22.0000
161   Default   1543.000000      characters   1543.000000  21.0000  21.0000
13    Default   1198.000000             123   1198.000000  20.0000  20.0000
238   Default   1314.000000            help   1314.000000  19.0000  19.0000
157   Default   1072.000000  capitalization   1072.000000  18.0000  18.0000
195   Default   1072.000000           doesn   1072.000000  17.0000  17.0000
234   Default    842.000000           guess    842.000000  16.0000  16.0000
555   Default    610.000000        reversed    610.000000  15.0000  15.0000
236   Default    610.000000          harder    610.000000  14.0000  14.0000
119   Default    610.000000            aren    610.000000  13.0000  13.0000
682   Default    910.000000         weakest    910.000000  12.0000  12.0000
205   Default    238.000000            easy    238.000000  11.0000  11.0000
651   Default    232.000000       uppercase    232.000000  10.0000  10.0000
373   Default    232.000000       lowercase    232.000000   9.0000   9.0000
263   Default    202.000000             jan    202.000000   8.0000   8.0000
17    Default    176.000000          123456    176.000000   7.0000   7.0000
159   Default    176.000000             cek    176.000000   6.0000   6.0000
179   Default    165.000000           david    165.000000   5.0000   5.0000
407   Default    163.000000          martin    163.000000   4.0000   4.0000
239   Default    144.000000           heslo    144.000000   3.0000   3.0000
16    Default    118.000000           12345    118.000000   2.0000   2.0000
200   Default    246.000000             don    246.000000   1.0000   1.0000
...       ...           ...             ...           ...      ...      ...
159   Topic10    175.363653             cek    176.304266   2.9394  -3.1737
179   Topic10    164.112080           david    165.052697   2.9391  -3.2400
16    Topic10    117.498457           12345    118.439073   2.9368  -3.5742
473   Topic10     91.780603             nik     92.721213   2.9346  -3.8212
83    Topic10     80.529028             333     81.469644   2.9332  -3.9520
569   Topic10     75.706934           roman     76.647546   2.9324  -4.0137
48    Topic10     74.099568            1986     75.040180   2.9322  -4.0352
248   Topic10     70.884838             hov     71.825448   2.9316  -4.0795
53    Topic10     62.848002            1991     63.788614   2.9299  -4.1999
57    Topic10     59.633276            1995     60.573883   2.9291  -4.2524
55    Topic10     56.418530            19

## Deep exploration

In [112]:
def pswd_analyse(results):
    mylist = []
    for i in range(len(results['sequence'])):
        for key, val in results['sequence'][i].items():
            if key == 'pattern':
                if val == 'dictionary':
                    #if results['sequence'][i]['dictionary_name'] != 'passwords':
                    mylist.append(results['sequence'][i]['dictionary_name'])
                elif val == 'sequence':
                    mylist.append(results['sequence'][i]['sequence_name'])
                elif val == 'regex':
                    mylist.append(results['sequence'][i]['regex_name'])
                elif val == 'date':
                    mylist.append(val)
                elif val == 'bruteforce':
                    mylist.append(val)
                elif val == 'repeat':
                    mylist.append(val)
                elif val == 'spatial':
                    mylist.append(val)
                else:
                    mylist.append(val)
                    print(val)
                    print(results)
    return(mylist)

In [184]:
# test
results = zxcvbn('HelloWorld')
print(results)
mylist = pswd_analyse(results)
mylist

{'password': 'HelloWorld', 'guesses': Decimal('45192'), 'guesses_log10': 4.655061561728271, 'sequence': [{'pattern': 'dictionary', 'i': 0, 'j': 4, 'token': 'Hello', 'matched_word': 'hello', 'rank': 83, 'dictionary_name': 'passwords', 'reversed': False, 'l33t': False, 'base_guesses': 83, 'uppercase_variations': 2, 'l33t_variations': 1, 'guesses': 166, 'guesses_log10': 2.220108088040055}, {'pattern': 'dictionary', 'i': 5, 'j': 9, 'token': 'World', 'matched_word': 'world', 'rank': 53, 'dictionary_name': 'english_wikipedia', 'reversed': False, 'l33t': False, 'base_guesses': 53, 'uppercase_variations': 2, 'l33t_variations': 1, 'guesses': 106, 'guesses_log10': 2.02530586526477}], 'calc_time': datetime.timedelta(microseconds=879), 'crack_times_seconds': {'online_throttling_100_per_hour': Decimal('1626912.000000000090311758072'), 'online_no_throttling_10_per_second': Decimal('4519.2'), 'offline_slow_hashing_1e4_per_second': Decimal('4.5192'), 'offline_fast_hashing_1e10_per_second': Decimal('0.

['passwords', 'english_wikipedia']

In [128]:
import re

data['withoutspecialwords'] = ''
for index, row in data.iterrows():
    # simple split for each row
    row['withoutspecialwords'] = "".join(re.findall(r'[0-9]+|[a-zA-Z]+',row['original']))
    

Take about 20s

In [152]:
data['pswd_analysis'] = ''
for index, row in data.iterrows():
    if row['withoutspecialwords']:
        results = zxcvbn(row['withoutspecialwords'])
        mylist = pswd_analyse(results)

    
    row['pswd_analysis'] = " ".join(mylist)
   # data['fixcap_with_pswd_strength'] = 
    

In [179]:
data['feedback'][43]

"Add another word or two. Uncommon words are better. Capitalization doesn't help very much."

In [180]:
docs_raw = list(data['fixcapital']+" "+data['password_strength']+" "+data['feedback']+" "+data['pswd_analysis'])
print(len(docs_raw))

38650


In [171]:
def LDA_pwstrenth(docs_raw,n_topics):
    tf_vectorizer = CountVectorizer(strip_accents = 'unicode',
                                    stop_words = ['english','aren','doesn','this','very','don','like','much','than','help','thems'],
                                    lowercase = True,
                                    # token choose the words with more than 4 length and get rid of the speical words
                                    token_pattern = r'\b[a-zA-Z0-9]{3,}\b',
                                    max_df = 0.5, 
                                    min_df = 10)
    dtm_tf = tf_vectorizer.fit_transform(docs_raw)
    tfidf_vectorizer = TfidfVectorizer(**tf_vectorizer.get_params())
    dtm_tfidf = tfidf_vectorizer.fit_transform(docs_raw)
    # for TF DTM
    lda_tf = LatentDirichletAllocation(n_components=n_topics, random_state=0)
    %time lda_tf.fit(dtm_tf)
    # for TFIDF DTM
    lda_tfidf = LatentDirichletAllocation(n_components=n_topics, random_state=0)
    %time lda_tfidf.fit(dtm_tfidf)
    return(lda_tf, dtm_tf, tf_vectorizer)




In [181]:
# Take about 2 min
# lda_tf_passwd strenth plus version
lda_tf_pwplus, dtm_tf_pwplus, tf_vectorizer_pwplus = LDA_words_numbers(docs_raw,5)
pyLDAvis.sklearn.prepare(lda_tf_pwplus, dtm_tf_pwplus, tf_vectorizer_pwplus)

/Users/xihajun/anaconda3/lib/python3.7/site-packages/sklearn/feature_extraction/text.py:1567: UserWarning: Only (<class 'numpy.float64'>, <class 'numpy.float32'>, <class 'numpy.float16'>) 'dtype' should be used. <class 'numpy.int64'> 'dtype' will be converted to np.float64.
  UserWarning)


CPU times: user 34.3 s, sys: 23.6 ms, total: 34.3 s
Wall time: 34.3 s
CPU times: user 28.3 s, sys: 22.3 ms, total: 28.3 s
Wall time: 28.3 s


/Users/xihajun/anaconda3/lib/python3.7/site-packages/pyLDAvis/_prepare.py:257: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  return pd.concat([default_term_info] + list(topic_dfs))


PreparedData(topic_coordinates=              x         y  topics  cluster       Freq
topic                                                
3     -0.253469  0.077901       1        1  25.573222
0      0.251782  0.267919       2        1  25.017641
4     -0.017301 -0.109287       3        1  20.788820
1      0.288381 -0.203170       4        1  14.576812
2     -0.269392 -0.033363       5        1  14.043505, topic_info=     Category          Freq            Term         Total  loglift  logprob
term                                                                       
626   Default  10879.000000          strong  10879.000000  30.0000  30.0000
178   Default   6015.000000            date   6015.000000  29.0000  29.0000
689   Default  11696.000000            weak  11696.000000  28.0000  28.0000
557   Default   3885.000000          repeat   3885.000000  27.0000  27.0000
627   Default   3473.000000       strongest   3473.000000  26.0000  26.0000
697   Default   3820.000000           years   3820.000000  25.0000  25.0000
129   Default   5596.000000           avoid   5596.000000  24.0000  24.0000
124   Default   3661.000000      associated   3661.000000  23.0000  23.0000
179   Default   3502.000000           dates   3502.000000  22.0000  22.0000
423   Default   8836.000000          medium   8836.000000  21.0000  21.0000
506   Default   6448.000000       passwords   6448.000000  20.0000  20.0000
161   Default   1629.000000      characters   1629.000000  19.0000  19.0000
558   Default   1629.000000        repeated   1629.000000  18.0000  18.0000
630   Default   1967.000000        surnames   1967.000000  17.0000  17.0000
690   Default   1154.000000         weakest   1154.000000  16.0000  16.0000
240   Default   1147.000000            help   1147.000000  15.0000  15.0000
157   Default    927.000000  capitalization    927.000000  14.0000  14.0000
197   Default    927.000000           doesn    927.000000  13.0000  13.0000
375   Default    559.000000           lower    559.000000  12.0000  12.0000
236   Default    590.000000           guess    590.000000  11.0000  11.0000
119   Default    428.000000            aren    428.000000  10.0000  10.0000
238   Default    428.000000          harder    428.000000   9.0000   9.0000
560   Default    428.000000        reversed    428.000000   8.0000   8.0000
193   Default   2094.000000          digits   2094.000000   7.0000   7.0000
13    Default    956.000000             123    956.000000   6.0000   6.0000
618   Default    407.000000         spatial    407.000000   5.0000   5.0000
92    Default    182.000000             666    182.000000   4.0000   4.0000
241   Default    151.000000           heslo    151.000000   3.0000   3.0000
650   Default    150.000000         torrent    150.000000   2.0000   2.0000
202   Default    209.000000             don    209.000000   1.0000   1.0000
...       ...           ...             ...           ...      ...      ...
18     Topic5     72.105879       123456789     72.923301   1.9517  -5.2752
365    Topic5     69.860934            link     70.677114   1.9514  -5.3068
367    Topic5     63.109173             lol     63.938727   1.9500  -5.4084
398    Topic5     60.874743             mar     61.691477   1.9497  -5.4445
215    Topic5     59.752521            enka     60.568213   1.9495  -5.4631
475    Topic5     56.382360            nick     57.198487   1.9486  -5.5211
222    Topic5     55.259967           filip     56.075498   1.9484  -5.5412
250    Topic5     49.643735             hov     50.459425   1.9467  -5.6484
333    Topic5     48.521459            kova     49.336237   1.9464  -5.6713
605    Topic5     47.397806             sky     48.212917   1.9460  -5.6947
354    Topic5     46.273415           lenka     47.089746   1.9455  -5.7187
374    Topic5     45.148711            love     45.966490   1.9451  -5.7433
5      Topic5     42.902263             100     43.720021   1.9441  -5.7944
556    Topic5     40.658443            rent     41.473681   1.9432  -5.8481
611    

In [149]:
data['original'][609]

'íííí'